torch/backends/_nnapi/serializer.py

```diff
@@ -0,0 +1,1131 @@
+        cache_key = (code, value)
+        if cache_key not in self.cached_immediates:
+            operand_id = len(self.operands)
+            self.operands.append(Operand(code, dims, DimOrder.SCALAR_OR_VECTOR, 0.0, 0))
+            self.values.append((operand_id, OperandValueSourceType.IMMEDIATE))
+            self.value_data.append(value)
+            self.cached_immediates[cache_key] = operand_id
+        return self.cached_immediates[cache_key]
+
+
+    def add_immediate_int_scalar(self, value):
+        return self.add_immediate_operand(
+                NNAPI_OperandCode.INT32,
+                struct.pack("i", value),
+                ())
+
+
+    def add_immediate_float_scalar(self, value):
+        return self.add_immediate_operand(
+                NNAPI_OperandCode.FLOAT32,
+                struct.pack("f", value),
+                ())
+
+
+    def add_immediate_bool_scalar(self, value):
+        return self.add_immediate_operand(
+                NNAPI_OperandCode.BOOL,
+                b"\x01" if value else b"\x00",
+                ())
+
+
+    def get_tensor_operand_by_jitval(self, jitval):
+        operand_id = self.jitval_operand_map[jitval]
+        return (operand_id, self.operands[operand_id])
+
+
+    def get_tensor_operand_or_constant(self, jitval):
+        operand_id = self.jitval_operand_map.get(jitval)
+        if operand_id is None:
+            _, value = self.get_constant_value(jitval, "TensorType")
+            operand_id = self.add_tensor_operand_for_weight(value)
+        return (operand_id, self.operands[operand_id])
+
+
+    def get_tensor_operand_for_weight(self, jitval):
+        _, value = self.get_constant_value(jitval, "TensorType")
+        operand_id = self.add_tensor_operand_for_weight(value)
+        return (operand_id, self.operands[operand_id])
+
+
+    def add_operation(self, opcode, inputs, outputs):
+        self.operations.append((opcode, len(inputs), len(outputs)))
+        self.operation_args.extend(inputs + outputs)
+
+    def add_constant_value(self, jitval, ctype, value):
+        assert jitval not in self.constants
+        self.constants[jitval] = (ctype, value)
+
+    def get_constant_value(self, jitval, typekind=None):
+        record = self.constants.get(jitval)
+        if record is None:
+            raise Exception(f"Could not find constant value for '{jitval!r}'.")
+        ctype, _ = record
+        if typekind is not None and ctype.kind() != typekind:
+            raise Exception(f"Expected constant value of type {typekind}, but got {ctype.kind()} for value '{jitval!r}'")
+        return record
+
+    def get_size_arg(self, jitval):
+        ctype, value = self.get_constant_value(jitval)
+        if ctype.kind() == "ListType":
+            assert ctype.getElementType().kind() == "IntType"
+            return value
+        raise Exception(f"Can't handle size arg of type '{ctype!r}' for '{jitval!r}'")
+
+    def expand_sizes(self, size):
+        return [ s.item() for s in size ]
```

Comment:
This method seems to be unused.

Reply:
Probably left-over from the old Caffe2 implementation.